source/extensions/filters/common/ratelimit/ratelimit_impl.cc

```diff
@@ -0,0 +1,145 @@
+      envoy::api::v2::ratelimit::RateLimitDescriptor::Entry* new_entry =
+          new_descriptor->add_entries();
+      new_entry->set_key(entry.key_);
+      new_entry->set_value(entry.value_);
+    }
+  }
+}
+
+void GrpcTlsClientImpl::limit(RequestCallbacks& callbacks, const std::string& domain,
+                              const std::vector<Envoy::RateLimit::Descriptor>& descriptors,
+                              Tracing::Span& parent_span) {
+  ASSERT(callbacks_ == nullptr);
+  callbacks_ = &callbacks;
+
+  envoy::service::ratelimit::v2::RateLimitRequest request;
+  createRequest(request, domain, descriptors);
+  request_ =
+      async_client_->send(service_method_, request, *this, parent_span, shared_state_->timeout_ms_);
+}
+
+void GrpcTlsClientImpl::onSuccess(
+    std::unique_ptr<envoy::service::ratelimit::v2::RateLimitResponse>&& response,
+    Tracing::Span& span) {
+  LimitStatus status = LimitStatus::OK;
+  ASSERT(response->overall_code() != envoy::service::ratelimit::v2::RateLimitResponse_Code_UNKNOWN);
+  if (response->overall_code() ==
+      envoy::service::ratelimit::v2::RateLimitResponse_Code_OVER_LIMIT) {
+    status = LimitStatus::OverLimit;
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceOverLimit);
+  } else {
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceOk);
+  }
+
+  Http::HeaderMapPtr headers = std::make_unique<Http::HeaderMapImpl>();
+  if (response->headers_size()) {
+    for (const auto& h : response->headers()) {
+      headers->addCopy(Http::LowerCaseString(h.key()), h.value());
+    }
+  }
+  callbacks_->complete(status, std::move(headers));
+  callbacks_ = nullptr;
+}
+
+void GrpcTlsClientImpl::onFailure(Grpc::Status::GrpcStatus status, const std::string&,
+                                  Tracing::Span&) {
+  ASSERT(status != Grpc::Status::GrpcStatus::Ok);
+  callbacks_->complete(LimitStatus::Error, nullptr);
+  callbacks_ = nullptr;
+}
+
+GrpcFactoryImpl::GrpcFactoryImpl(const envoy::config::ratelimit::v2::RateLimitServiceConfig& config,
+                                 Grpc::AsyncClientManager& async_client_manager,
+                                 ThreadLocal::SlotAllocator& tls, Stats::Scope& scope)
+    : tls_(tls) {
+  envoy::api::v2::core::GrpcService grpc_service;
+  grpc_service.MergeFrom(config.grpc_service());
+  // TODO(htuch): cluster_name is deprecated, remove after 1.6.0.
+  if (config.service_specifier_case() ==
+      envoy::config::ratelimit::v2::RateLimitServiceConfig::kClusterName) {
+    grpc_service.mutable_envoy_grpc()->set_cluster_name(config.cluster_name());
+  }
+  async_client_factory_ = async_client_manager.factoryForGrpcService(grpc_service, scope, false);
+}
+
+ClientPtr GrpcFactoryImpl::create(const absl::optional<std::chrono::milliseconds>& timeout,
+                                  Server::Configuration::FactoryContext& context) {
+  return context.singletonManager().getTyped<GrpcClientImpl>(
```

Comment:
@mattklein123 Here is where I am storing client in singleton manager and this will remain as is in future also. We do not have to change this part. That is why I am not moving this to ratelimit_registration.cc. My plan is that ratelimit_registration will completely go away when we finally delete the rate limit config from bootstrap.